File: candil-api/src/main/java/com/velas/candil/services/CandleService.java
package com.velas.candil.services;

import com.velas.candil.models.candle.CategoryEnum;
import com.velas.candil.models.candle.FeatureEnum;
import com.velas.candil.models.candle.MaterialEnum;
import com.velas.candil.models.candle.CandleRequestDto;
import com.velas.candil.models.candle.CandleResponseDto;
import com.velas.candil.models.candle.CandleUpdateDto;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.Date;

public interface CandleService extends CrudService<CandleRequestDto, CandleResponseDto, CandleUpdateDto, Long> {

    Page<CandleResponseDto> findByCategory(Pageable pageable, CategoryEnum categoryEnum);
    Page<CandleResponseDto> findByMaterial(Pageable pageable, MaterialEnum materialEnum);
    Page<CandleResponseDto> findByFeature(Pageable pageable, FeatureEnum featureEnum);
    Page<CandleResponseDto> findByDate(Pageable pageable, Date date);
    Page<CandleResponseDto> findByNameContaining(Pageable pageable, String name);
}

File: candil-ui/src/app/core/services/candle.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import {
  CandleRequest, CandleUpdateRequest, CandleResponse,
  CategoryEnum, MaterialEnum, FeatureEnum,
  Page
} from '../../shared/models/candle.models';

@Injectable({ providedIn: 'root' })
export class CandleService {
  private readonly http = inject(HttpClient);
  private readonly base = 'http://localhost:8080/v1/candles';

  findAll(page = 0, size = 12): Observable<Page<CandleResponse>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<Page<CandleResponse>>(this.base, { params });
  }

  findById(id: number): Observable<CandleResponse> {
    return this.http.get<CandleResponse>(`${this.base}/${id}`);
  }

  create(dto: CandleRequest): Observable<CandleResponse> {
    return this.http.post<CandleResponse>(this.base, dto);
  }

  update(id: number, dto: CandleUpdateRequest): Observable<CandleResponse> {
    return this.http.put<CandleResponse>(`${this.base}/${id}`, dto);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.base}/${id}`);
  }

  findByCategory(category: CategoryEnum, page = 0, size = 12): Observable<Page<CandleResponse>> {
    const params = new HttpParams()
      .set('categoryEnum', category)
      .set('page', page)
      .set('size', size);
    return this.http.get<Page<CandleResponse>>(`${this.base}/category`, { params });
  }

  findByMaterial(material: MaterialEnum, page = 0, size = 12): Observable<Page<CandleResponse>> {
    const params = new HttpParams()
      .set('materialEnum', material)
      .set('page', page)
      .set('size', size);
    return this.http.get<Page<CandleResponse>>(`${this.base}/material`, { params });
  }

  findByFeature(feature: FeatureEnum, page = 0, size = 12): Observable<Page<CandleResponse>> {
    const params = new HttpParams()
      .set('featureEnum', feature)
      .set('page', page)
      .set('size', size);
    return this.http.get<Page<CandleResponse>>(`${this.base}/feature`, { params });
  }

  search(name: string, page = 0, size = 12): Observable<Page<CandleResponse>> {
    const params = new HttpParams()
      .set('name', name)
      .set('page', page)
      .set('size', size);
    return this.http.get<Page<CandleResponse>>(`${this.base}/search`, { params });
  }
}